### ISSUE TITLE
Enso developers should be able to generate reports about profiling results of top-level sections of the application.

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4276]
This task is automatically imported from the old Task Issue Board and it was originally created by Michael Mauderer.
Original issue is [here](https://www.pivotaltracker.com/story/show/180311774).
<hr />

## Team
Technical Designer: KW
Implementer: KW
Acceptance/QA: MM

## Why
Having the profiling framework in place, we should connect it to the application to see real data and understand the current performance bottlenecks.

**As** a developer
**I want** to generate a performance report showing information about top level events/section level (e.g., "loading WASM" / "opening project" ).
**So that** I can easily spot performance issues or set up automated performance checks. 

## Acceptance Criteria

```gherkin
Scenario: 
Given I have the Enso application
When  I run the application with  the command line argument `--profile`
Then a performance report is generated and stored at `temp/enso/profiling/<date-with-time>.json`  and it contains information about top-level events in the application.
```

## Resources
https://github.com/enso-org/design/blob/wip/wd/profiling/designs/profiling/implementation.md#batch-results

## Tasks:
- [X] Set up the profiling points to measure the hierarchical processes described in the document.
- [X] Discover any other processes not described in the document where profiling should be attached to, so there are not time regions not covered by the profiling.
- [X] Implement the new command to generate the report.

## Blockers:
[#181192616](https://www.pivotaltracker.com/story/show/#181192616) resolved

## Comments:
**Michael Mauderer** reports a new **STANDUP** for yesterday (2021-12-20):

**Today:** Worked on revisions of the initial profiling PR. Some planning on finishing the profiling epic. 1on1 with team and quick check-in meeting. It should be finished by 2021-12-22.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Chatting with Wojciech  about some details of the API macro implementations, planning and getting re-review. Getting properly started on setting up the profiling in the codebase. (Enso Bot - Dec 21, 2021)
<hr />
**Michael Mauderer** reports a new **STANDUP** for today (2021-12-21):

**Today:** Worked on revisions of the initial profiling PR. Go started on adding profiling sections to the codebase. It should be finished by 2021-12-22.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Continuing work on defining sections for profiling in the codebase. (Enso Bot - Dec 21, 2021)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for yesterday (2021-12-22):

**Summary:** There is **1 day** delay in implementation of the [Enso developers should be able to generate reports about profiling results of top-level sections of the application. (#180311774) task](https://www.pivotaltracker.com/story/show/180311774).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Revisions to the previous PR took more time than expected due to unforeseen issues with the requested API changes. (Enso Bot - Dec 23, 2021)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2021-12-22):

**Today:** Worked on adding profiling sections to the codebase and noticed some issues in the original API that needed fixing. Waiting on feedback on previous PR. It should be finished by 2021-12-23.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Continuing work on defining sections for profiling in the codebase. (Enso Bot - Dec 23, 2021)
<hr />
**Michael Mauderer** reports a new **🔴 DELAY** for yesterday (2022-01-12):

**Summary:** There is **25 day** delay in implementation of the [Enso developers should be able to generate reports about profiling results of top-level sections of the application. (#180311774) task](https://www.pivotaltracker.com/story/show/180311774).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Not a real delay as such, but I meant to get started on this task just before Christmas, which ended up not happening due to finishing another PR having precedence. So, I'm now continuing after the holidays and re-planning. (Enso Bot - Jan 13, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-01-12):

**Today:** More drafting and coming up with the initial task for Kaz. Not really got much done on the new task, but will get going the next day. It should be finished by 2022-01-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Getting started on the next task. (Enso Bot - Jan 13, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for yesterday (2022-01-13):

**Today:** Got back into the code I started before Christmas and made some updates and extensions to cover some more parts of the codebase with profiling information. Noticed some thins that could make this process easier and started drafting some suggestions for API updates. It should be finished by 2022-01-17.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Continue work on task, post profiling API suggestion updates, PR reviews, and 1on1 meetings with most of the IDE team. (Enso Bot - Jan 14, 2022)
<hr />
**Michael Mauderer** reports a new **STANDUP** for the provided date (2022-01-14):

**Today:** Spent my time on 1on1s and discussions about re-planning of the performance profiling framework and the performance logging of EnsoGL stats. It should be finished by 2022-01-17.

**Next Day:** Next day I will be working on the [##180868978 task](https://www.pivotaltracker.com/story/show/180868978). Summarize 1on1 feedback and pass it on, continue re-planning of profiling framework with feedback from Kaz and EnsoGL data collection with Mateusz. This will pause the implementation task until all re-plannig is properly done. (Enso Bot - Jan 17, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for today (2022-02-28):

**Summary:** There is **44 days** delay in implementation of the [Enso developers should be able to generate reports about profiling results of top-level sections of the application. (#180311774) task](https://www.pivotaltracker.com/story/show/180311774).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Issue was started with original profiler library, went back to unstarted while we redesigned that; today I'm starting an implementation using the new profiler API. (Enso Bot - Feb 28, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-02-28):

**Today:** Started on #180311774. It should be finished by 2022-03-02.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Debug a rustc segfault. Continue adding high-level profiling to app. (Enso Bot - Feb 28, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-03-01):

**Today:** Continuing on #180311774. Solved some problems blocking collecting real app data. It should be finished by 2022-03-02.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Continue adding high-level profiling to app. (Enso Bot - Mar 1, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for today (2022-03-02):

**Summary:** There is **2 days** delay in implementation of the [Enso developers should be able to generate reports about profiling results of top-level sections of the application. (#180311774) task](https://www.pivotaltracker.com/story/show/180311774).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Needed some changes to the low-level API so that its data meets the same State-transition invariants as the high-level API. (Enso Bot - Mar 2, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-03-02):

**Today:** Continuing on #180311774: now collecting enough data to shed light on where time is spent on startup! Required some changes to the low-level API so that it produces data that meets the same State-transition invariants as the high-level API. It should be finished by 2022-03-04.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Connect to graph renderer? (Enso Bot - Mar 2, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-03-03):

**Today:** Instrumentation is working; implemented generating reports in a textual format. It should be finished by 2022-03-04.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Connect to graph renderer. (Enso Bot - Mar 3, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for the last Friday (2022-03-04):

**Today:** Rendered some real data. Found we needed to be able to distinguish idle time from time doing unprofiled work; implemented that. It should be finished by 2022-03-04.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Polish PR for delivery. (Enso Bot - Mar 7, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for yesterday (2022-03-07):

**Summary:** There is **3 days** delay in implementation of the [Enso developers should be able to generate reports about profiling results of top-level sections of the application. (#180311774) task](https://www.pivotaltracker.com/story/show/180311774).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** A lot of little things came up when gathering real data and connecting to graphs (Enso Bot - Mar 8, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-03-07):

**Today:** Wrapping up #180311774. It should be finished by 2022-03-07.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180311774). Start on more detailed instrumentation. (Enso Bot - Mar 8, 2022)
<hr />
**Acceptance/QA** review set to **pass** (Michael Mauderer - Apr 6, 2022)
<hr />
